Allow configuration of Feign timeout and our client's maximum backoff.

harrybiddle opened the pull request:
<!---
Please remember to:
- Assign someone to review this PR
- Add any necessary release notes (including breaking changes)
- Make sure the documentation is up to date for your change
--->

I need to be able to configure the Feign timeout and turn off exponential backoff in our client, in order to be able to make Jepsen fast.

Not sure how to do this: I've put up this PR as a strawman, but very open to hear neater suggestions!!

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/atlasdb/1296)
<!-- Reviewable:end -->

jeremyk-91 commented on atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java:
```diff
@@ -94,6 +95,23 @@ private AtlasDbHttpClients() {
                 .target(failoverFeignTarget);
     }
 
+    public static <T> T createProxyWithQuickFailover(
```
I'd suggest you can name this as `createProxyWithFailover()` (still with the 5 arguments), and have the above factory call it. The `3000` is a constant from the `FailoverFeignTarget` that perhaps could be made public.

The 4th argument could be an `Optional<Request.Options>` maybe even - then we could apply options to the builder only if the optional is present, perhaps?

Then you'd have  `createProxyWithFailover(sslSocketFactory, endpointUris, 3000, Optional.absent(), type);` in the other method. (There might be an even better way to do this too!)

harrybiddle commented:
I've passed the `Request.Options` directly in rather than making in an `Optional`. If you don't want to override these options, you simply use the default constructor.

jeremyk-91 commented on atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java:
```diff
@@ -38,6 +38,8 @@
+    public static final int DEFAULT_MAX_BACKOFF_MILLIS = 3000;
+
     private static final Logger log = LoggerFactory.getLogger(FailoverFeignTarget.class);
     private static final double GOLDEN_RATIO = (Math.sqrt(5) + 1.0) / 2.0;
```
For consistency we normally have the logger as separate from everything else (I'm not sure why this wasn't originally in the target). Could we reorder these to

```
public class FFT<T> implements ... {
  psf Logger log = ...;

  psf int DMBM = 3000;
  (etc.)
}

jeremyk-91 commented on atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java:
```diff
@@ -88,7 +89,18 @@ private AtlasDbHttpClients() {
+
+    /**
+     * As above, but you can optionally override Feign options, and the maximum time that the failover client will
+     * backoff for.
```
nit: I'd prefer for us to not use "as above", because Javadoc has a rather interesting way of sorting its methods (it sorts them alphabetically and then by types). Also IntelliJ's autocompletion (Ctrl-J for me) has its own ordering too, so it may be confusing to see "as above" without the context.

SerialVelocity commented on atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java:
```diff
@@ -56,8 +59,13 @@
+    public FailoverFeignTarget(Collection<String> servers, int maxBackoffMillis, Class<T> type) {
         this.servers = ImmutableList.copyOf(ImmutableSet.copyOf(servers));
         this.type = type;
+        this.maxBackoffMillis = maxBackoffMillis;
```
We should fail fast if this is 0 or below

harrybiddle replied:
Done - now throwing an exception

SerialVelocity commented on atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java:
```diff
@@ -88,7 +89,18 @@ private AtlasDbHttpClients() {
     public static <T> T createProxyWithFailover(
             Optional<SSLSocketFactory> sslSocketFactory, Collection<String> endpointUris, Class<T> type) {
-        FailoverFeignTarget<T> failoverFeignTarget = new FailoverFeignTarget<>(endpointUris, type);
+        return createProxyWithFailover(sslSocketFactory, endpointUris, new Request.Options(),
```
Can we put all of these on new lines?

harrybiddle replied:
Done

harrybiddle commented:
Changes since last review:

- Throw `IllegalArgumentException` if constructing a `FailoverFeignTarget` with a non-positive `maxBackoffMillis`
- Moved the `createProxyWithFailover` that takes optional Feign timeout and max backoff _out_ of the public API.
- Instead, Jepsen can use a package-level `createProxyWithFailoverWithoutBackoff` method, that's annotated with `@VisibleForTesting` to make it clear that this should not be used by users of AtlasDB.

SerialVelocity commented on atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java:
```diff
@@ -41,6 +43,8 @@
 
     private static final int CONNECTION_POOL_SIZE = 100;
     private static final long KEEP_ALIVE_TIME_MILLIS = TimeUnit.MILLISECONDS.convert(10, TimeUnit.MINUTES);
+    private static final int ONE_MIILISECOND = 1;
```
`MIILI`

harrybiddle replied:
Done!

SerialVelocity commented on atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java:
```diff
@@ -56,8 +59,16 @@
+    }
+
+    public FailoverFeignTarget(Collection<String> servers, int maxBackoffMillis, Class<T> type) {
+        if (maxBackoffMillis <= 0) {
```
`Preconditions.checkArgument`

SerialVelocity replied:
It also supports `{}` for substitution

harrybiddle replied:
Nice, that's much better.  Switched it out

jeremyk-91 commented on atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java:
```diff
@@ -97,9 +117,22 @@ private AtlasDbHttpClients() {
+    @VisibleForTesting
+    static <T> T createProxyWithFailoverWithoutBackoff(
+            Optional<SSLSocketFactory> sslSocketFactory, Collection<String> endpointUris, Class<T> type) {
+        Request.Options options = new Request.Options(ONE_MIILISECOND, ONE_MIILISECOND);
```
Hmm. I was concerned that this was dangerous - the Feign options are actually _timeouts_, not really backoff related. We cut them down to stop the tests stalling for 10 or 60s on a doomed request, but this seems a bit excessive.

But the tests are passing, so I must be mistaken here. I'm not sure the name is representative of what the method's doing though.

harrybiddle replied:
It's a fair point, we might be able to play with this a little bit to get more out of the tests. I think that's a bit too involved for now though, since the tests do seem to be working and giving us signal.

I agree with you about the name! Struggled a little here; I've gone for `createProxyWithQuickFailoverForTesting`. What do you think?
